massa-ledger-worker/src/ledger_db.rs

```diff
@@ -463,37 +464,43 @@ impl LedgerDB {
     /// * The last taken key (this is an optimization to easily keep a reference to the last key)
     pub fn get_ledger_part(
         &self,
-        last_key: &Option<Vec<u8>>,
-    ) -> Result<(Vec<u8>, Option<Vec<u8>>), ModelsError> {
-        let ser = VecU8Serializer::new();
-        let key_serializer = KeySerializer::new();
+        cursor: StreamingStep<Vec<u8>>,
+    ) -> Result<(Vec<u8>, StreamingStep<Vec<u8>>), ModelsError> {
         let handle = self.db.cf_handle(LEDGER_CF).expect(CF_ERROR);
-        let mut part = Vec::new();
         let opt = ReadOptions::default();
+        let ser = VecU8Serializer::new();
+        let key_serializer = KeySerializer::new();
+        let mut ledger_part = Vec::new();
 
         // Creates an iterator from the next element after the last if defined, otherwise initialize it at the first key of the ledger.
-        let db_iterator = if let Some(key) = last_key {
-            let mut iter =
-                self.db
-                    .iterator_cf_opt(handle, opt, IteratorMode::From(key, Direction::Forward));
-            iter.next();
-            iter
-        } else {
-            self.db.iterator_cf_opt(handle, opt, IteratorMode::Start)
+        let (db_iterator, mut new_cursor) = match cursor {
+            StreamingStep::Started => (
+                self.db.iterator_cf_opt(handle, opt, IteratorMode::Start),
+                StreamingStep::Started,
+            ),
+            StreamingStep::Ongoing(last_key) => {
+                let mut iter = self.db.iterator_cf_opt(
+                    handle,
+                    opt,
+                    IteratorMode::From(&last_key, Direction::Forward),
+                );
+                iter.next();
+                (iter, StreamingStep::Finished)
+            }
+            StreamingStep::Finished => return Ok((ledger_part, cursor)),
         };
-        let mut last_key = None;
 
         // Iterates over the whole database
         for (key, entry) in db_iterator.flatten() {
-            if (part.len() as u64) < (self.ledger_part_size_message_bytes) {
-                key_serializer.serialize(&key.to_vec(), &mut part)?;
-                ser.serialize(&entry.to_vec(), &mut part)?;
-                last_key = Some(key.to_vec());
+            if (ledger_part.len() as u64) < (self.ledger_part_size_message_bytes) {
+                key_serializer.serialize(&key.to_vec(), &mut ledger_part)?;
+                ser.serialize(&entry.to_vec(), &mut ledger_part)?;
+                new_cursor = StreamingStep::Ongoing(key.to_vec());
             } else {
                 break;
             }
         }
-        Ok((part, last_key))
+        Ok((ledger_part, new_cursor))
```

Comment:
This cursor don't reach finished when all data has been sent

Reply:
It does because it starts at `Finished` [line 488](https://github.com/massalabs/massa/pull/3147/files#diff-1219a2cf735a0b9e73b4372b43c8191d263af08fae67b6e13eb30e855f2eca61R488)